c-dependencies/js-compute-runtime/builtins/object-store.cpp

```diff
@@ -0,0 +1,464 @@
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  auto acme_challenge = ".well-known/acme-challenge/";
+  if (strncmp(key_chars, acme_challenge, strlen(acme_challenge)) == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_KEY_ACME);
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  if (strcmp(key_chars, ".") == 0 || strcmp(key_chars, "..") == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_KEY_RELATIVE);
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  JS::HandleValue body_val = args.get(1);
+
+  // We currently support five types of body inputs:
+  // - byte sequence
+  // - buffer source
+  // - USV strings
+  // - URLSearchParams
+  // - ReadableStream
+  // After the other other options are checked explicitly, all other inputs are
+  // encoded to a UTF8 string to be treated as a USV string.
+  // TODO: Support the other possible inputs to Body.
+
+  JS::RootedObject body_obj(cx, body_val.isObject() ? &body_val.toObject() : nullptr);
+
+  if (body_obj && JS::IsReadableStream(body_obj)) {
+    if (RequestOrResponse::body_unusable(cx, body_obj)) {
+      JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                                JSMSG_READABLE_STREAM_LOCKED_OR_DISTRUBED);
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+
+    // If the body stream is backed by a C@E body handle, we can directly pipe
+    // that handle into the object store.
+    if (NativeStreamSource::stream_is_body(cx, body_obj)) {
+      JS::RootedObject stream_source(cx, NativeStreamSource::get_stream_source(cx, body_obj));
+      JS::RootedObject source_owner(cx, NativeStreamSource::owner(stream_source));
+      BodyHandle body = RequestOrResponse::body_handle(source_owner);
+
+      int status = fastly_object_store_insert(object_store_handle(self), key_chars, key_len, body);
+      if (!HANDLE_RESULT(cx, status)) {
+        return ReturnPromiseRejectedWithPendingError(cx, args);
+      }
+
+      // The insert was successful so we return a Promise which resolves to undefined
+      JS::RootedValue rval(cx);
+      rval.setUndefined();
+      JS::ResolvePromise(cx, result_promise, rval);
+      args.rval().setObject(*result_promise);
+
+      return true;
+    } else {
+      JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                                JSMSG_OBJECT_STORE_PUT_CONTENT_STREAM);
+      return ReturnPromiseRejectedWithPendingError(cx, args);
+    }
+  } else {
+    mozilla::Maybe<JS::AutoCheckCannotGC> maybeNoGC;
+    JS::UniqueChars text;
+    char *buf;
+    size_t length;
+
+    if (body_obj && JS_IsArrayBufferViewObject(body_obj)) {
+      // Short typed arrays have inline data which can move on GC, so assert
+      // that no GC happens. (Which it doesn't, because we're not allocating
+      // before `buf` goes out of scope.)
```

Comment:
It might be worth being explicit about `maybeNoGC` needing to be populated for the lifetime of `buf` in this case, otherwise it might be tempting to have it clean itself up at the end of this block.